types/scene/subsurface_tree.c

```diff
@@ -0,0 +1,215 @@
+#include <assert.h>
+#include <stdlib.h>
+#include <wlr/types/wlr_scene.h>
+#include <wlr/util/addon.h>
+
+/**
+ * A tree for a surface and all of its child sub-surfaces.
+ *
+ * `tree` contains `scene_surface` and one node per sub-surface.
+ */
+struct wlr_scene_subsurface_tree {
+	struct wlr_scene_tree *tree;
+	struct wlr_surface *surface;
+	struct wlr_scene_surface *scene_surface;
+
+	struct wlr_scene_subsurface_tree *parent; // NULL for the top-level surface
+	struct wlr_addon surface_addon; // only set if there's a parent
+
+	struct wl_listener tree_destroy;
+	struct wl_listener surface_destroy;
+	struct wl_listener surface_commit;
+	struct wl_listener surface_new_subsurface;
+};
+
+static void surface_tree_destroy(struct wlr_scene_subsurface_tree *surface_tree) {
```

Comment:
I think this file could use a `s/surface_tree/subsurface_tree/g` after the rename of the struct.